packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol

```diff
@@ -0,0 +1,92 @@
+    // The only L2 contract that can send messages over the bridge via the messenger is the oracle spoke.
+    address public oracleSpoke;
+
+    // Messenger contract on the other side of the L1<->L2 bridge.
+    address public parentMessenger;
+
+    // Hard coded default gas limit for L1 transactions.
+    uint32 public defaultGasLimit = 5_000_000;
+
+    // TODO: import from optimism contracts when they release their latest version.
+    address internal constant L2_CROSS_DOMAIN_MESSENGER = 0x4200000000000000000000000000000000000007;
+
+    /**
+     * @notice Construct the Optimism_ChildMessenger contract.
+     * @param _parentMessenger The address of the L1 parent messenger. Acts as the "owner" of this contract.
+     */
+    constructor(address _parentMessenger) OVM_CrossDomainEnabled(L2_CROSS_DOMAIN_MESSENGER) {
+        parentMessenger = _parentMessenger;
+    }
+
+    /**
+     * @notice Changes the stored address of the Oracle spoke, deployed on L2.
+     * @dev The caller of this function must be the parent messenger, over the canonical bridge.
+     * @param newOracleSpoke address of the new oracle spoke, deployed on L2.
+     */
+    function setOracleSpoke(address newOracleSpoke) public onlyFromCrossDomainAccount(parentMessenger) {
+        oracleSpoke = newOracleSpoke;
+    }
+
+    /**
+     * @notice Changes the stored address of the parent messenger, deployed on L1.
+     * @dev The caller of this function must be the parent messenger, over the canonical bridge.
+     * @param newParentMessenger address of the new parent messenger, deployed on L1.
+     */
+    function setParentMessenger(address newParentMessenger) public onlyFromCrossDomainAccount(parentMessenger) {
+        parentMessenger = newParentMessenger;
+    }
+
+    /**
+     * @notice Changes the default gas limit that is sent along with transactions to Ethereum.
+     * @dev The caller of this function must be the parent messenger, over the canonical bridge.
+     * @param newDefaultGasLimit the new L1 gas limit to be set.
+     */
+    function setDefaultGasLimit(uint32 newDefaultGasLimit) public onlyFromCrossDomainAccount(parentMessenger) {
+        defaultGasLimit = newDefaultGasLimit;
+    }
+
+    /**
+     * @notice Sends a message to the parent messenger via the canonical message bridge.
+     * @dev The caller must be the OracleSpoke on L2. No other contract is permissioned to call this function.
+     * @dev The L1 target, the parent messenger, must implement processMessageFromChild to consume the message.
+     * @param data data message sent to the L1 messenger. Should be an encoded function call or packed data.
+     */
+    function sendMessageToParent(bytes memory data) public override {
+        require(msg.sender == oracleSpoke, "Only callable by oracleSpoke");
+        sendCrossDomainMessage(
+            parentMessenger,
+            defaultGasLimit,
+            abi.encodeWithSignature("processMessageFromChild(bytes)", data)
```

Comment:
Thoughts on changing this to `abi.encode(block.chainId, abi.encodeWithSignature("processMessageFromChild(bytes)", data))`?

Then the parent can `abi.decode(data, (uint256, bytes))` and call `oracleHub.processMessageFromChild(chainId, data)`.

This way the OracleHub doesn't hardcode the child messenger's chain ID, but simply passes on whatever chain ID the child messenger detects from its environment?

Or, the ParentMessenger can `require(decodedChainId == chainId)` to enforce one way communication between ParentMessenger and a ChildMessenger for the `chainId` network.

Reply:
I think validating this is somewhat redundant as the hardcoding in `onlyFromCrossDomainAccount` enforces who the `childMessenger` is;  it is true that there could be multiple chains with the same ID on this but this is simply restricted via the `_crossDomainMessenger`, which is configured to only work between two chains (eg ethereum and optimism).

Regarding changing the data structure so the oracleHub does not need to store the Id I also think makes things more complex than they should be; it is important that you can look at the ParentMessanger and see which chain this is pointing to by reading the state variable `childChainId`.

Regarding the extra encoding, I think this makes this really hard to reason about without much gain.

Overall, I think it best to _not_ make this change for the reasons listed above ^^

Reply:
> but this is simply restricted via the _crossDomainMessenger, which is configured to only work between two chains (eg ethereum and optimism).

This is a good point, can ignore